mplc/multi_partner_learning.py

```diff
@@ -663,6 +664,724 @@ def fit_minibatch(self):
+                    model.compiled_metrics.update_state(y, y_pred)  # log the loss and accuracy
+                    partners_optimizers[p_id].minimize(loss, model.trainable_weights,
+                                                       tape=tape)  # perform local optimization
+
+                for model_w, partner_w in zip(model.trainable_weights,
+                                              partners_weights[p_id]):  # update the partner's weights
+                    partner_w.assign(model_w.read_value())
+            # at the end of the minibatch, aggregate all the local weights
+            aggregated_weights = aggregate_weights(tuple(partners_weights),
+                                                   agg_w)
+
+            # inform the partners with the new weights
+            for p_id in range(len(partners_minibatches)):
+                for new_w, partner_w in zip(aggregated_weights, partners_weights[p_id]):
+                    partner_w.assign(new_w.read_value())
+
+        @tf.function
+        def fit_minibatch(model, partners_minibatches, partners_optimizers, partners_weights, agg_w, smodel_list):
+            for p_id, minibatch in enumerate(partners_minibatches):  # minibatch == (x,y)
+                # minibatch[0] in a tensor of shape=(number of batch, batch size, img).
+                # We cannot iterate on tensors, so we convert this tensor to a list of
+                # *number of batch* tensors with shape=(batch size, img)
+                x_minibatch = tf.unstack(minibatch[0], axis=0)
+                y_minibatch = tf.unstack(minibatch[1], axis=0)  # same here, with labels
+
+                for model_w, partner_w in zip(model.trainable_weights,
+                                              partners_weights[p_id]):  # set the model weights to partner's ones
+                    model_w.assign(partner_w.read_value())
+
+                for x, y in zip(x_minibatch, y_minibatch):  # iterate over batches
+                    s_model_p = smodel_list[p_id]
+                    with tf.GradientTape() as tape:
+                        yt_pred = model(x)
+                        yo_pred = s_model_p(yt_pred)
+                        loss = model.compiled_loss(y, yo_pred)
+                    s_model_p.optimizer.minimize(loss, s_model_p.trainable_weights, tape=tape)
+                    partners_optimizers[p_id].minimize(loss, model.trainable_weights,
+                                                       tape=tape)
+                    model.compiled_metrics.update_state(y, yo_pred)
+
+                for model_w, partner_w in zip(model.trainable_weights,
+                                              partners_weights[p_id]):  # update the partner's weights
+                    partner_w.assign(model_w.read_value())
+            # at the end of the minibatch, aggregate all the local weights
+            aggregated_weights = aggregate_weights(tuple(partners_weights),
```

Comment:
This code is very similar to the `fit_minibactch()` of the other class above. Could it be factorized?

Reply:
I don't think so, because fit_minibatch is a tf.function, so we need to re-declare it every time we start a new fit. 
Because a tf.function can initialize variables only on its first call, and we do need fit_minibatches to initialize things (optimizer, model weights, etc) when call fit.